### PR TITLE
[clang-query] add basic profiling on matching each ASTs

### PR DESCRIPTION
**Pull request by danix800**
We've found that basic profiling could help improving/optimizing when developing clang-tidy checks.

This PR adds an extra command
```
set enable-profile (true|false)   Set whether to enable matcher profiling.
```
which enables profiling queries on each file.

Sample output:

```
$ cat test.cql
set enable-profile true
m binaryOperator(isExpansionInMainFile())

$ cat test.c
int test(int i, int j) {
  return i + j;
}

$ clang-query --track-memory -f test.cql test.c --

Match #1:

{{.*}}/test.c:2:10: note: "root" binds here
    2 |   return i + j;
      |          ^~~~~
1 match.
===-------------------------------------------------------------------------===
                         clang-query matcher profiling
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0000 seconds (0.0000 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  ---Mem---  --- Name ---
   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)        224  {{.*}}/test.c
   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)        224  Total
```

### CODE REVIEW

File: clang-tools-extra/clang-query/QueryProfile.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+namespace clang::query {
+
+class QueryProfile {
+public:
+  llvm::StringMap<llvm::TimeRecord> Records;
+  QueryProfile() = default;
+  ~QueryProfile();
+
+private:
+  std::optional<llvm::TimerGroup> TG;
```
Comment (AaronBallman) [inline clang-tools-extra/clang-query/QueryProfile.h:29]: Why is this `optional`?

File: clang-tools-extra/clang-query/QueryProfile.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace llvm {
+class raw_ostream;
+} // namespace llvm
+
+namespace clang::query {
+
+class QueryProfile {
+public:
+  llvm::StringMap<llvm::TimeRecord> Records;
```
Comment (AaronBallman) [inline clang-tools-extra/clang-query/QueryProfile.h:24]: Rather than leave this public, should we add an interface to insert a new record?

**Review by AaronBallman — COMMENTED** [review]
This is really cool, thank you! Can you also add a release note to `clang-tools-extra/docs/ReleaseNotes.rst` so users know about the new functionality?

Changes generally look good, I just had a few questions.

File: clang-tools-extra/clang-query/QueryProfile.h (hunk 1 above)
Comment (danix800) [reply clang-tools-extra/clang-query/QueryProfile.h:29]: No need to be optional, I'll remove it.

File: clang-tools-extra/clang-query/QueryProfile.h (hunk 2 above)
Comment (danix800) [reply clang-tools-extra/clang-query/QueryProfile.h:24]: `QueryProfile` is a very small utility like class, I'll move it into Query.cpp.

Comment (danix800) [reply clang-tools-extra/clang-query/QueryProfile.h]: QueryProfiler could be simplified to no more then 10 lines.

**Comment by danix800** [comment]
> This is really cool, thank you! Can you also add a release note to `clang-tools-extra/docs/ReleaseNotes.rst` so users know about the new functionality?
> 
> Changes generally look good, I just had a few questions.

ReleaseNotes updated.

**Review by AaronBallman — APPROVED** [review]
LGTM, thank you!

**Pull request merged**